### ISSUE TITLE
[Defactor] Support Map-with-masking

### ISSUE THREAD

**Issue by midfield** [genjax-community/genjax#632]
Essentially, it would be nice to have a "length" argument to Map akin to the Unfold combinator, that would expose VectorTraceType length property to Map users.  The use case is, often in JAX one wants to work with statically-sized batches of things but pass around a "length" parameter which says how many things you actually have.  Then Map could do the masking without user intervention.

cc @brianwa84

**Comment by femtomc** [comment]
@midfield I can write some documentation about how to use the masking system, but it will allow you to express this pattern, with a bit of user code.

**Comment by midfield** [comment]
As I write more of the DCOLMAP generative model, the need for general masking primitives becomes more and more pressing, as we need to deal with a dynamic number of objects / keypoints in many places.  I have some hacks that work in basic cases, but the general problem is not solved.

**Comment by femtomc** [comment]
@midfield do you want to meet next Tuesday (can't be Monday, unfortunately, as I have a midterm) to discuss exactly what you need? I can quickly identify if the current system will work for your use case, and can finish writing the documentation for it.

**Comment by midfield** [comment]
Happy to meet, but what I need is pretty simple to describe: a `MaskedMap` combinator that works exactly the same as the `Map` combinator, but results in a function that takes an extra validity mask argument (probably first argument) in addition to the usual function arguments.  The validity mask is a boolean vector with the same size as the `in_axes` batch dimension, and specifies which elements of the batch to ignore for scoring purposes.  (In my use case, the validity mask will always look like [True, True, ..., False, False].)  Basically, expose the `Unfold` combinator masking to `Map`.

This is just one API, I'd be happy with any other API which exposes the same functionality (vectorize but ignore some of the indexes for scoring purposes) will work for me.

**Comment by derifatives** [comment]
@midfield, just to clarify, do you need an arbitrary subset of indices or do you need "the first k indices?"

**Comment by femtomc** [comment]
@midfield Here's a self-contained blurb which works on `nightly`.

Now, I won't promise this will be stable -- there's a few (small) things here which will be affected by some of other defactor plans, but can you play with this? I'm just showing you `simulate` and `importance` (and also, roughly, how to use `genjax.mask` and the `Mask.match` interface).

```python
from dataclasses import dataclass
import genjax
from genjax import Trace
from genjax import GenerativeFunction, JAXGenerativeFunction, MapCombinator, SupportsBuiltinSugar, mask
from genjax.typing import Tuple, BoolArray, FloatArray, Any
import jax
import jax.numpy as jnp

console = genjax.pretty()

@dataclass
class MaskedMapTrace(Trace):
    gen_fn: GenerativeFunction
    mask: BoolArray
    inner: Trace
    args: Tuple
    retval: Any
    score: FloatArray

    def flatten(self):
        return (
            self.gen_fn,
            self.mask,
            self.inner,
            self.args,
            self.retval,
            self.score,
        ), ()

    def get_score(self):
        return self.get_score()

    def get_gen_fn(self):
        return self.gen_fn

    def get_args(self):
        return self.args

    def get_retval(self):
        return self.retval

    def get_score(self):
        return self.score


@dataclass
class MaskedMapCombinator(JAXGenerativeFunction, SupportsBuiltinSugar):
    map_combinator: MapCombinator

    def flatten(self):
        return (self.map_combinator,), ()

    def simulate(self, key, args):
        mask_array, *normal_args = args
        tr = self.map_combinator.simulate(key, tuple(normal_args))
        inner_tr = tr.inner
        score = jnp.sum(jnp.multiply(inner_tr.get_score(), mask_array))
        retval = tr.get_retval()
        masked_retval = mask(mask_array, retval)
        return MaskedMapTrace(
            self, mask_array, inner_tr, normal_args, masked_retval, score
        )

    def importance(self, key, chm, args):
        mask_array, *normal_args = args
        masked_chm = mask(mask_array, chm)
        w, tr = self.map_combinator.importance(key, masked_chm, tuple(normal_args))
        inner_tr = tr.inner
        score = jnp.sum(jnp.multiply(inner_tr.get_score(), mask_array))
        retval = tr.get_retval()
        masked_retval = mask(mask_array, retval)
        return w, MaskedMapTrace(
            self, mask_array, inner_tr, normal_args, masked_retval, score
        )

    
@genjax.gen
def model(x):
    return genjax.tfp_normal(x, 1.0) @ "x"

mcomb = genjax.map_combinator(model, in_axes=(0, ))
m_mcomb = MaskedMapCombinator(mcomb)

key = jax.random.PRNGKey(314159)
tr = jax.jit(m_mcomb.simulate)(key, (jnp.array([True, False]), jnp.array([1.0, 2.0])))

retval = tr.get_retval()
console.print((tr.get_score(), tr.get_retval()))
console.print((retval.mask, retval.value))

# Mask matching interface.
retval.match(lambda: jnp.array([0.0, 0.0]), lambda v: v)
```

**Comment by midfield** [comment]
@derifatives I just need the last k to be masked.

@femtomc Thanks for the code.  I'm a little bit wary about introducing brittleness into DCOLMAP, but I'll give it a go.

**Comment by sritchie** [comment]
Let us know if this is still an issue after @femtomc 's help!
